File: addons/guide-cpp/src/inputs/guide_input_joy_button.cpp
#include "guide_input_joy_button.h"
#include "guide_input_state.h"

using namespace godot;

void GUIDEInputJoyButton::_bind_methods() {
    ClassDB::bind_method(D_METHOD("get_button"), &GUIDEInputJoyButton::get_button);
    ClassDB::bind_method(D_METHOD("set_button", "button"), &GUIDEInputJoyButton::set_button);
    ADD_PROPERTY(PropertyInfo(Variant::INT, "button"), "set_button", "get_button");

    ClassDB::bind_method(D_METHOD("_refresh"), &GUIDEInputJoyButton::_refresh);
    ClassDB::bind_method(D_METHOD("_to_string"), &GUIDEInputJoyButton::_to_string);
}

GUIDEInputJoyButton::GUIDEInputJoyButton() {
}

GUIDEInputJoyButton::~GUIDEInputJoyButton() {
}

void GUIDEInputJoyButton::_begin_usage() {
    if (_state) {
        _state->connect("joy_button_state_changed", Callable(this, "_refresh"));
        _refresh();
    }
}

void GUIDEInputJoyButton::_end_usage() {
    if (_state) {
        _state->disconnect("joy_button_state_changed", Callable(this, "_refresh"));
    }
}

void GUIDEInputJoyButton::_refresh() {
    if (_state) {
        _value.x = _state->is_joy_button_pressed(joy_index, (JoyButton)button) ? 1.0 : 0.0;
    }
}

bool GUIDEInputJoyButton::is_same_as(const Ref<GUIDEInput> &other) const {
    Ref<GUIDEInputJoyButton> o = other;
    if (o.is_null()) return false;
    return o->button == button && o->joy_index == joy_index;
}

String GUIDEInputJoyButton::_to_string() const {
    return "(GUIDEInputJoyButton: button=" + String::num(button) + ", joy_index=" + String::num(joy_index) + ")";
}

String GUIDEInputJoyButton::_editor_name() const {
    return "Joy Button";
}

String GUIDEInputJoyButton::_editor_description() const {
    return "A button press from a joy button.";
}

GUIDEAction::GUIDEActionValueType GUIDEInputJoyButton::_native_value_type() const {
    return GUIDEAction::BOOL;
}

File: addons/GuideCS/Inputs/GUIDEInputJoyButton.cs
using System;
using Godot;

namespace Guide;

[Tool]
[GlobalClass]
public partial class GUIDEInputJoyButton : GUIDEInputJoyBase {
	private JoyButton _button = JoyButton.A;
	[Export]
	public JoyButton Button {
		get => _button;
		set {
			if (_button == value) return;
			_button = value;
			EmitChanged();
		}
	}

	public override void _BeginUsage() {
		_State.JoyButtonStateChanged += _Refresh;
		_Refresh();
	}

	public override void _EndUsage() {
		_State.JoyButtonStateChanged -= _Refresh;
	}

	private void _Refresh() {
		_Value = new Vector3(_State.IsJoyButtonPressed(JoyIndex, Button) ? 1.0f : 0.0f, 0, 0);
	}

	public override bool IsSameAs(GUIDEInput other) {
		return other is GUIDEInputJoyButton joyButton &&
			   joyButton.Button == Button &&
			   joyButton.JoyIndex == JoyIndex;
	}

	public override string ToString() => $"(GUIDEInputJoyButton: button={Button}, joy_index={JoyIndex})";

	public override string _EditorName() => "Joy Button";

	public override string _EditorDescription() => "A button press from a joy button.";

	public override GUIDEAction.GUIDEActionValueType _NativeValueType() => GUIDEAction.GUIDEActionValueType.Bool;
}
